### ISSUE TITLE
Haskell generic bash completions break cross-compilation

### ISSUE THREAD

**Issue by roberth** [NixOS/nixpkgs#49748]
## Issue description

Completions in Haskell currently break cross compilation.

Since #49477 we have a function that does this generically for Haskell packages. This can be improved by either:
 - not adding completions when cross compiling
 - emulating the behavior of optparse-applicative to generate the very simple scripts in plain bash, which seems feasible. The duplicated logic can then be tested against the native optparse-applicative behavior.

### Steps to reproduce

Cross compile a package with completions, like stack. It will call the build output and fail.
Note that stack currently has its own override in hackage2nix, which is probably best to disable in favor of the library functions.

**Comment by roberth** [comment]
cc @Ericson2314

**Comment by Ericson2314** [comment]
I would just not do it during cross compilation for. K
Now. We can also copy the output from a native build.

**Comment by matthewbauer** [comment]
I think disabling makes the most sense. You could use the native-built one but it might end up being incorrect. For instance one option is only available for one architecture or operating system.

**Comment by roberth** [comment]
It's actually a very very predictable script that does not depend on the options. For this reason, duplicating it in Nix is actually feasible.

It depends only on the executable name/path and the command name.
[Source.](https://github.com/pcapriotti/optparse-applicative/blob/c38c20fb91713e1e8e1114b6fad1ce30d483da1c/Options/Applicative/BashCompletion.hs#L162)

Example:

```
$ nix-shell -p dhall --run 'stack --bash-completion-script dhall'
_stack()
{
    local CMDLINE
    local IFS=$'\n'
    CMDLINE=(--bash-completion-index $COMP_CWORD)

    for arg in ${COMP_WORDS[@]}; do
        CMDLINE=(${CMDLINE[@]} --bash-completion-word $arg)
    done

    COMPREPLY=( $(dhall "${CMDLINE[@]}") )
}

complete -o filenames -F _stack stack
```

**Comment by roberth** [comment]
Of course this is a 'nice to have'. We can do without completions on cross.

**Comment by roberth** [comment]
The protocol between shell and executable has been fairly stable and the maintainer cares about compatibility, because non-Nix users can have scripts that are out of sync with (upgraded) executables.
Additionally, the protocol may be upgraded to a more forward-compatible form similar to, say, http response headers. See https://github.com/pcapriotti/optparse-applicative/pull/414

**Comment by marksisson** [comment]
Is there a recommended way to disable the completions for cross compiling?

**Comment by roberth** [comment]
None of the overrides in nixpkgs seem to handle this.

You could define a function with the same name in the package set `pkgs/development/haskell-modules/make-package-set.nix`, and make its definition conditional so that it returns an empty string when cross.
Instead of `lib.generateOptparseApplicativeCompletion`, you'll call `super.generateOptparseApplicativeCompletion`.

**Comment by Artturin** [comment]
```diff
diff --git a/pkgs/development/haskell-modules/lib/compose.nix b/pkgs/development/haskell-modules/lib/compose.nix
index 600bf80cb19..575b2e08f11 100644
--- a/pkgs/development/haskell-modules/lib/compose.nix
+++ b/pkgs/development/haskell-modules/lib/compose.nix
@@ -437,9 +437,9 @@ rec {
       zshCompDir="''${!outputBin}/share/zsh/vendor-completions"
       fishCompDir="''${!outputBin}/share/fish/vendor_completions.d"
       mkdir -p "$bashCompDir" "$zshCompDir" "$fishCompDir"
-      "''${!outputBin}/bin/${exeName}" --bash-completion-script "''${!outputBin}/bin/${exeName}" >"$bashCompDir/${exeName}"
-      "''${!outputBin}/bin/${exeName}" --zsh-completion-script "''${!outputBin}/bin/${exeName}" >"$zshCompDir/_${exeName}"
-      "''${!outputBin}/bin/${exeName}" --fish-completion-script "''${!outputBin}/bin/${exeName}" >"$fishCompDir/${exeName}.fish"
+      ${pkgs.stdenv.hostPlatform.emulator pkgs.buildPackages} "''${!outputBin}/bin/${exeName}" --bash-completion-script "''${!outputBin}/bin/${exeName}" >"$bashCompDir/${exeName}"
+      ${pkgs.stdenv.hostPlatform.emulator pkgs.buildPackages} "''${!outputBin}/bin/${exeName}" --zsh-completion-script "''${!outputBin}/bin/${exeName}" >"$zshCompDir/_${exeName}"
+      ${pkgs.stdenv.hostPlatform.emulator pkgs.buildPackages} "''${!outputBin}/bin/${exeName}" --fish-completion-script "''${!outputBin}/bin/${exeName}" >"$fishCompDir/${exeName}.fish"
 
       # Sanity check
       grep -F ${exeName} <$bashCompDir/${exeName} >/dev/null || {
```

this would probably work, did not test

**Comment by marksisson** [comment]
Resolved by #ac1f1ad0e0
